# Here in this code, i have a dataset that contain data about many features of tumors, and i will be making a machine learning pipeline for predicting if cancer is malginent or benign, to help doctors in their tests

# and will then load ot for production

First i will upload the libraries we will need to upload the dataset and split the data and test and train the data

In [1]:
import pandas as pd
import sklearn.metrics
import sklearn.model_selection

then i will upload the dataset

knowing that there are many columns, i will use 'display.max_columns' option to display all of them and look for any non numeric or categorial data, if any non numeric exist i must encode it

In [2]:
pd.set_option('display.max_columns', None)
df = pd.read_csv("Cancer_Data.csv")
df

,id,diagnosis,radius_mean,texture_mean,perimeter_mean,area_mean,smoothness_mean,compactness_mean,concavity_mean,concave points_mean,symmetry_mean,fractal_dimension_mean,radius_se,texture_se,perimeter_se,area_se,smoothness_se,compactness_se,concavity_se,concave points_se,symmetry_se,fractal_dimension_se,radius_worst,texture_worst,perimeter_worst,area_worst,smoothness_worst,compactness_worst,concavity_worst,concave points_worst,symmetry_worst,fractal_dimension_worst,Unnamed: 32
0,842302,M,17.99,10.38,122.80,1001.0,0.11840,0.27760,0.30010,0.14710,0.2419,0.07871,1.0950,0.9053,8.589,153.40,0.006399,0.04904,0.05373,0.01587,0.03003,0.006193,25.380,17.33,184.60,2019.0,0.16220,0.66560,0.7119,0.2654,0.4601,0.11890,NaN
1,842517,M,20.57,17.77,132.90,1326.0,0.08474,0.07864,0.08690,0.07017,0.1812,0.05667,0.5435,0.7339,3.398,74.08,0.005225,0.01308,0.01860,0.01340,0.01389,0.003532,24.990,23.41,158.80,1956.0,0.12380,0.18660,0.2416,0.1860,0.2750,0.08902,NaN
2,84300903,M,19.69,21.25,130.00,1203.0,0.10960,0.15990,0.19740,0.12790,0.2069,0.05999,0.7456,0.7869,4.585,94.03,0.006150,0.04006,0.03832,0.02058,0.02250,0.004571,23.570,25.53,152.50,1709.0,0.14440,0.42450,0.4504,0.2430,0.3613,0.08758,NaN
3,84348301,M,11.42,20.38,77.58,386.1,0.14250,0.28390,0.24140,0.10520,0.2597,0.09744,0.4956,1.1560,3.445,27.23,0.009110,0.07458,0.05661,0.01867,0.05963,0.009208,14.910,26.50,98.87,567.7,0.20980,0.86630,0.6869,0.2575,0.6638,0.17300,NaN
4,84358402,M,20.29,14.34,135.10,1297.0,0.10030,0.13280,0.19800,0.10430,0.1809,0.05883,0.7572,0.7813,5.438,94.44,0.011490,0.02461,0.05688,0.01885,0.01756,0.005115,22.540,16.67,152.20,1575.0,0.13740,0.20500,0.4000,0.1625,0.2364,0.07678,NaN
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
564,926424,M,21.56,22.39,142.00,1479.0,0.11100,0.11590,0.24390,0.13890,0.1726,0.05623,1.1760,1.2560,7.673,158.70,0.010300,0.02891,0.05198,0.02454,0.01114,0.004239,25.450,26.40,166.10,2027.0,0.14100,0.21130,0.4107,0.2216,0.2060,0.07115,NaN
565,926682,M,20.13,28.25,131.20,1261.0,0.09780,0.10340,0.14400,0.09791,0.1752,0.05533,0.7655,2.4630,5.203,99.04,0.005769,0.02423,0.03950,0.01678,0.01898,0.002498,23.690,38.25,155.00,1731.0,0.11660,0.19220,0.3215,0.1628,0.2572,0.06637,NaN
566,926954,M,16.60,28.08,108.30,858.1,0.08455,0.10230,0.09251,0.05302,0.1590,0.05648,0.4564,1.0750,3.425,48.55,0.005903,0.03731,0.04730,0.01557,0.01318,0.003892,18.980,34.12,126.70,1124.0,0.11390,0.30940,0.3403,0.1418,0.2218,0.07820,NaN
567,927241,M,20.60,29.33,140.10,1265.0,0.11780,0.27700,0.35140,0.15200,0.2397,0.07016,0.7260,1.5950,5.772,86.22,0.006522,0.06158,0.07117,0.01664,0.02324,0.006185,25.740,39.42,184.60,1821.0,0.16500,0.86810,0.9387,0.2650,0.4087,0.12400,NaN


the values i saw for the last column (Unnamed: 32) are all NaN, so i will check if all of them are like this, if their is similar to the number of rows then i will delete the column

In [3]:
dfnull = df['Unnamed: 32'].isna().sum()
dfnull

np.int64(569)

569 is equal to the number of rows, so i will delete the column and also the id column because it's not needed in the training and don't offer any information.

In [4]:
df = df.drop(['id', 'Unnamed: 32'], axis=1)

Here i want to check if all the values in the diagnosis column are malignent or benign (no corrupted data in the column), i will count how many times both values are mentioned and then see it's equal to the number of total rows

In [5]:
C1 = (df["diagnosis"] == 'M').sum()
C2 = (df["diagnosis"] == 'B').sum()
print(C1)
print(C2)

212
357


then i will turn them to 1 and 0, so that they could be used for the trainig and testing of models

In [6]:
df['diagnosis'] = df['diagnosis'] .replace({'B': 1, 'M': 0})

/tmp/ipython-input-2509303161.py:1: FutureWarning: Downcasting behavior in `replace` is deprecated and will be removed in a future version. To retain the old behavior, explicitly call `result.infer_objects(copy=False)`. To opt-in to the future behavior, set `pd.set_option('future.no_silent_downcasting', True)`
  df['diagnosis'] = df['diagnosis'] .replace({'B': 1, 'M': 0})


Here i will split the dataset to train and test sets, and since the data isn't very large, i will take less data for test than 20%

In [7]:
df_train, df_test = sklearn.model_selection.train_test_split(df,test_size=0.15)
print("df:", df.shape)
print("df_train:", df_train.shape)
print("df_test:", df_test.shape)

df: (569, 31)
df_train: (483, 31)
df_test: (86, 31)


# Preprocess

First i will check for empty values to remove them in the training set

In [8]:
dfnull = df_train.isna().sum()
dfnull

,0
diagnosis,0
radius_mean,0
texture_mean,0
perimeter_mean,0
area_mean,0
smoothness_mean,0
compactness_mean,0
concavity_mean,0
concave points_mean,0
symmetry_mean,0


then i will check for duplicate rows, and i will delete if any exists

In [9]:
dfduplicate =df_train.duplicated().sum()
dfduplicate

np.int64(0)

Here i will check for outliers to delete, the dataset isn't very big so i will try to delete less number of the data

but here i will only pick the columns with big differece between their min and max values.

first i will check the max and min values in them, then remove the outliers

In [10]:
minval = df_train.min()
maxval = df_train.max()
print(minval)
print(maxval)

diagnosis                    0.000000
radius_mean                  6.981000
texture_mean                 9.710000
perimeter_mean              43.790000
area_mean                  143.500000
smoothness_mean              0.062510
compactness_mean             0.019380
concavity_mean               0.000000
concave points_mean          0.000000
symmetry_mean                0.106000
fractal_dimension_mean       0.049960
radius_se                    0.111500
texture_se                   0.360200
perimeter_se                 0.757000
area_se                      6.802000
smoothness_se                0.001713
compactness_se               0.002252
concavity_se                 0.000000
concave points_se            0.000000
symmetry_se                  0.007882
fractal_dimension_se         0.000895
radius_worst                 7.930000
texture_worst               12.020000
perimeter_worst             50.410000
area_worst                 185.200000
smoothness_worst             0.071170
compactness_

I saw that the columns that have biggest differences between their min amd max values are [area_mean,area_se,area_worst], so i will only remove the outlier values in them

In [11]:
Q1 = df_train['area_mean'].quantile(0.10)
Q3 = df_train['area_mean'].quantile(0.90)
IQR = Q3 - Q1

df_train = df_train[(df_train['area_mean'] >= Q1 - 1.5 * IQR) & (df_train['area_mean'] <= Q3 + 1.5 * IQR)]

In [12]:
Q1 = df_train['area_se'].quantile(0.10)
Q3 = df_train['area_se'].quantile(0.90)
IQR = Q3 - Q1

df_train = df_train[(df_train['area_se'] >= Q1 - 1.5 * IQR) & (df_train['area_se'] <= Q3 + 1.5 * IQR)]

In [13]:
Q1 = df_train['area_worst'].quantile(0.10)
Q3 = df_train['area_worst'].quantile(0.90)
IQR = Q3 - Q1

df_train = df_train[(df_train['area_worst'] >= Q1 - 1.5 * IQR) & (df_train['area_worst'] <= Q3 + 1.5 * IQR)]

In [14]:
df_train.shape

(479, 31)

and here i will split the train and test data for a sets for the the target label, and a sets for the other feautres

In [15]:
x_train = df_train.drop(["diagnosis"], axis = 1)
y_train = df_train["diagnosis"]

x_test = df_test.drop(["diagnosis"], axis = 1)
y_test = df_test["diagnosis"]

print("x train size:", x_train.shape)
print("x test size:", x_test.shape)
print("y train size:", y_train.shape)
print("y test size:", y_test.shape)

x train size: (479, 30)
x test size: (86, 30)
y train size: (479,)
y test size: (86,)


# Train, Evaluate and test

Here i will try to find the best model to train it and then test it
i must upload the libriaries i need for the training and testing of the model
I will use several ensemble learning methods and then pick the best using gridsearch

first i will upload the libraries needed for the
and ensemble learning methods

In [16]:
from sklearn.metrics import accuracy_score, f1_score, precision_score, recall_score, confusion_matrix
from sklearn.model_selection import GridSearchCV

the ensemble learning algorithms i picked are random forest for it's accuracy and handling complex relations between features, Extra trees because it's faster and the dataset have many dimensions, ADA because it's effective for binary classification, Voting classifier to use severla methods at once

In [17]:
from sklearn.ensemble import RandomForestClassifier
from sklearn.ensemble import ExtraTreesClassifier
from sklearn.ensemble import AdaBoostClassifier
from sklearn.ensemble import VotingClassifier

I will tune the hyperparameters for the Randomforestclassifer, i will set criterion for different ways to test impurity of splits, i will pick the default and a lower value for the n_estimator (since our dataset isn't very large)

In [18]:
RFC_parameters_grid = {"criterion" : ["gini", "entropy", "log_loss"],
                   "n_estimators": ( 50, 100),
                        "max_depth": (5,10,None)}
RFC_model = sklearn.model_selection.GridSearchCV(sklearn.ensemble.RandomForestClassifier(),
             RFC_parameters_grid, scoring="accuracy")
RFC_model.fit(x_train, y_train)
print ("accuracy {:.2f}".format(RFC_model.best_score_))
print ("best found hyperparameters  = {}".format(RFC_model.best_params_))

accuracy 0.97
best found hyperparameters  = {'criterion': 'gini', 'max_depth': None, 'n_estimators': 100}


for ada boosting classifier i will also pick some lower values with the default ones for learning rate and n_estimators, because the dataset isn't large, also different algoritms

In [19]:
ada_parameters_grid = { "learning_rate" : [ 0.1, 0.2, 1.0],
                   "n_estimators": (20, 50),
                   "algorithm": ('SAMME', 'SAMME.R')}
ADAmodel = sklearn.model_selection.GridSearchCV(sklearn.ensemble.AdaBoostClassifier(),
             ada_parameters_grid, scoring="accuracy")
ADAmodel.fit(x_train, y_train)
print ("accuracy {:.2f}".format(ADAmodel.best_score_))
print ("best found hyperparameters = {}".format(ADAmodel.best_params_))

/usr/local/lib/python3.12/dist-packages/sklearn/ensemble/_weight_boosting.py:519: FutureWarning: The parameter 'algorithm' is deprecated in 1.6 and has no effect. It will be removed in version 1.8.
  warnings.warn(
/usr/local/lib/python3.12/dist-packages/sklearn/ensemble/_weight_boosting.py:519: FutureWarning: The parameter 'algorithm' is deprecated in 1.6 and has no effect. It will be removed in version 1.8.
  warnings.warn(
/usr/local/lib/python3.12/dist-packages/sklearn/ensemble/_weight_boosting.py:519: FutureWarning: The parameter 'algorithm' is deprecated in 1.6 and has no effect. It will be removed in version 1.8.
  warnings.warn(
/usr/local/lib/python3.12/dist-packages/sklearn/ensemble/_weight_boosting.py:519: FutureWarning: The parameter 'algorithm' is deprecated in 1.6 and has no effect. It will be removed in version 1.8.
  warnings.warn(
/usr/local/lib/python3.12/dist-packages/sklearn/ensemble/_weight_boosting.py:519: FutureWarning: The parameter 'algorithm' is deprecated in 

accuracy 0.96
best found hyperparameters = {'algorithm': 'SAMME', 'learning_rate': 1.0, 'n_estimators': 50}


in Extra tree classifier, i also use default/lower values for n_estimators and max depth, several criterion methods,and raise minumum decrease of impurity needed to split a node and allow out of bag samples

In [20]:
ETC_parameters_grid = { "n_estimators": [50,100],
                       "criterion": ['gini', 'entropy'],
                        "max_depth": [None,5,10],
                       "min_impurity_decrease":[0.0, 0.5],
                        "bootstrap": [True],
                   "oob_score": (True, False)}
ETCmodel = sklearn.model_selection.GridSearchCV(sklearn.ensemble.ExtraTreesClassifier(),
             ETC_parameters_grid, scoring="accuracy")
ETCmodel.fit(x_train, y_train)
print ("accuracy {:.2f}".format(ETCmodel.best_score_))
print ("best found hyperparameters = {}".format(ETCmodel.best_params_))

accuracy 0.97
best found hyperparameters = {'bootstrap': True, 'criterion': 'gini', 'max_depth': None, 'min_impurity_decrease': 0.0, 'n_estimators': 50, 'oob_score': True}


for the voting classifier, i must first upload and use some classificaiton algorithms and tune them

For svm/svc, i will try 3 different kernels

for logistic regression i will try using all processors and or just one

for Multinomial naiv Bayes, i will use prior probabilities or not

for the decision tree, to measure quality of the tree spliting i will try gini and entropy
different tree depths, and numbers required for split to avoid overfitting
put some values for min_impurity_decrease to put more restriction for splits in the tree by putting minimum amount of impurity decrease to split a node

for the Voting Classifier i will use both hard and soft voting

In [21]:
from sklearn.tree import DecisionTreeClassifier
from sklearn.naive_bayes import MultinomialNB
from sklearn.svm import SVC
from sklearn.linear_model import LogisticRegression
from sklearn.ensemble import VotingClassifier
from sklearn.model_selection import GridSearchCV

# Define classifiers
T = DecisionTreeClassifier()
# setting probability as true for SVC will make it take into consideration the probality of the results of the classifcation
S = SVC(probability=True)
MNB = MultinomialNB()
# make a big border for max iteraitons to remove limits and any wanrings and to make reuslts as accurate as possible
L = LogisticRegression(max_iter=10000)

# Define parameter grids for each classifier
T_param_grid = {'criterion': ['gini', 'entropy'], 'max_depth': [3, 5, 10],
                'min_samples_split': [1, 2], 'min_impurity_decrease': [1, 2, 0.0]}
S_param_grid = {'kernel': ['linear', 'rbf', 'poly']}
L_param_grid = {'solver': ['lbfgs', 'liblinear', 'sag', 'saga'],
                'max_iter': [500, 1000, 2000]}
MNB_param_grid = {'fit_prior': [True, False]}

In [22]:
estimators = [('dt', T), ('svm', S), ('NB', MNB), ('LR', L)]

# Define parameter grid for Voting Classifier
VC_parameters_grid = {"voting": ['hard', 'soft']}

# Create Voting Classifier
voting_classifier = VotingClassifier(estimators)

# Perform Grid Search with cross-validation
VC_model = GridSearchCV(voting_classifier, VC_parameters_grid, scoring="accuracy")
VC_model.fit(x_train, y_train)

# Print the best accuracy and hyperparameters
print("Accuracy: {:.2f}".format(VC_model.best_score_))
print("Best hyperparameters:", VC_model.best_params_)

Accuracy: 0.95
Best hyperparameters: {'voting': 'hard'}


and here i want to pick the best model

In [23]:
Bestmodel =  RFC_model.best_estimator_ if RFC_model.best_score_ > ADAmodel.best_score_ and RFC_model.best_score_ > ETCmodel.best_score_ and RFC_model.best_score_ > VC_model.best_score_ else ADAmodel.best_estimator_ if ADAmodel.best_score_ > ETCmodel.best_score_ and ADAmodel.best_score_ > VC_model.best_score_ else ETCmodel.best_estimator_ if ETCmodel.best_score_ > VC_model.best_score_ else VC_model.best_estimator_
Bestmodel.fit(x_train, y_train)

ExtraTreesClassifier(bootstrap=True, n_estimators=50, oob_score=True)

and here i will see the result of the model on the test set

In [24]:
y_pred = Bestmodel.predict(x_test)
accuracy = accuracy_score(y_test, y_pred)
f1 = f1_score(y_test, y_pred)
precision = precision_score(y_test, y_pred)
recall = recall_score(y_test, y_pred)
CM = confusion_matrix(y_test, y_pred)

print("accuracy:", accuracy)
print("confusion matrix:\n", CM)
print("FI:", f1)
print("Precision:", precision)
print("recall:", recall)

accuracy: 0.9767441860465116
confusion matrix:
 [[41  2]
 [ 0 43]]
FI: 0.9772727272727273
Precision: 0.9555555555555556
recall: 1.0


# Saving the model to use it later

In [25]:
# importing needed library and saving the model
import joblib

joblib.dump(Bestmodel, "the_model.pkl")
print("\nModel saved as the_model.pkl")


Model saved as the_model.pkl


In [26]:
# example usage of the model
load_model = joblib.load("the_model.pkl")

example = x_test.iloc[[5]]
prediction = load_model.predict(example)
print(prediction)

[0]


# deploying the model with gradio

gradio is a python library that helps in deploying MLMs with a URL.

In [43]:
# install it
!pip install gradio

In [84]:
# importing and launching it
import gradio as gr

model = joblib.load("the_model.pkl")

columns = ["radius_mean", "texture_mean", "perimeter_mean","area_mean", "smoothness_mean","compactness_mean",
            "concavity_mean", "concave points_mean", "symmetry_mean", "fractal_dimension_mean", "radius_se", "texture_se",
            "perimeter_se", "area_se", "smoothness_se", "compactness_se", "concavity_se", "concave points_se", "symmetry_se",
            "fractal_dimension_se", "radius_worst", "texture_worst", "perimeter_worst", "area_worst", "smoothness_worst",
            "compactness_worst", "concavity_worst", "concave points_worst", "symmetry_worst", "fractal_dimension_worst"
            ]
# map positional inputs to columns names and convert to dataframe
def tumor_type_prediction(*args):
    try:
       features = dict(zip(columns, args))
       df = pd.DataFrame([features])
       df = df[model.feature_names_in_]
       # predicting
       prediction = model.predict(df)[0]
       return "The Tumor is Malignant" if prediction == 1 else "The Tumor is benign"
    # dealing with errors
    except Exception as e:
        return f"Prediction failed: {e}"
# building the interface of the URL
iface = gr.Interface(
    fn = tumor_type_prediction,
    inputs = [gr.Number(label=col) for col in columns],
    outputs = "text",
    title = "Tumors classifer app"
)
iface.launch(share=True)

Colab notebook detected. To show errors in colab notebook, set debug=True in launch()
* Running on public URL: https://230dafd18ed6f2570b.gradio.live

This share link expires in 1 week. For free permanent hosting and GPU upgrades, run `gradio deploy` from the terminal in the working directory to deploy to Hugging Face Spaces (https://huggingface.co/spaces)


Dataset refrencing:

Taha, E. (2023) Cancer data, Kaggle. Available at: https://www.kaggle.com/datasets/erdemtaha/cancer-data (Accessed: 27 June 2023).